src/reducers/console.js

```diff
@@ -35,20 +33,62 @@ export default function console(stateIn, {type, payload, meta}) {
         ),
       );
     case 'EVALUATE_CONSOLE_ENTRY':
-      return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
-        new ConsoleEntry({expression: payload}),
+      return payload.trim(' ') === '' ? state : state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({expression: payload}),
       );
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_CHANGE':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
-        meta.key,
-        new ConsoleEntry({
+      return state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
       );
+    case 'NAVIGATE_CONSOLE_HISTORY': {
+      const relevantHistory = state.history.toList().filter(
+        entry => entry.expression !== null,
+      );
+
+      const newHistoryEntryIndex = payload.direction === 'UP' ?
+        state.historyEntryIndex + 1 :
+        state.historyEntryIndex - 1;
+
+      if (newHistoryEntryIndex < 0 ||
+          newHistoryEntryIndex === relevantHistory.size) {
+        return state;
+      }
+
+      const downToPreviousInput = newHistoryEntryIndex === 0 &&
+        payload.direction === 'DOWN';
+
+      if (downToPreviousInput) {
+        return state.
+          delete('nextConsoleEntry').
+          delete('historyEntryIndex').
+          set('currentInputValue', state.nextConsoleEntry);
+      }
+
+      const historyIndex = relevantHistory.size - newHistoryEntryIndex;
+      const {expression} = relevantHistory.get(historyIndex);
+      const updatedState = state.
+        set('historyEntryIndex', newHistoryEntryIndex).
+        set('currentInputValue', expression);
+
+
+      const firstUp = newHistoryEntryIndex === 1 &&
+        payload.direction === 'UP';
+
+      if (firstUp) {
+        return updatedState.set('nextConsoleEntry', state.currentInputValue);
+      }
+
+      return updatedState;
+    }
```

Comment:
OK so this is definitely looking better, I do want to make one more pitch for always maintaining the `nextConsoleEntry` though!! In particular, I think it would allow you to remove these lines:

  ```js
const firstUp = newHistoryEntryIndex === 1 &&
  payload.direction === 'UP';

if (firstUp) {
  return updatedState.set('nextConsoleEntry', state.currentInputValue);
}
```

Which seems like a pretty big win, not as much in terms of literal lines of code removed as just dropping a special case and making the whole thing easier to reason about.

If you still don’t think it’s worth it though I will shut up about it : )

Reply:
For me, my mental model is this: "up and down arrows update the state related to history, and typing directly updates the current input value"

I'll admit that a conditional update, which necessitates the temporary variable updatedState, isn't great, but I think it's worth it to keep the CONSOLE_CHANGED handling as simple as

```
    case 'CONSOLE_CHANGE':
      return state.set('currentInputValue', payload.value);
```

Reply:
For reference, here's what it would look like to keep the next input value always in sync:

```js
case 'CONSOLE_CHANGE': {
  if (state.historyEntryIndex === 0) {
    return state.set('nextInputValue', payload.value).set('currentInputValue', payload.value)
  }
  return state.set('currentInputValue', payload.value);
}
```

What all this makes me wonder is if there's a way to apply an update unconditionally and apply an update conditionally without duplicated code or temporary variables. I guess helper functions could help

Reply:
The [`withMutations()`](https://facebook.github.io/immutable-js/docs/#/Map/withMutations) method might be a good way to apply an update conditionally without needing a temporary variable?